Improved handling of kwargs being passed to #initialize’s super method

timriley opened the pull request:
This PR adjusts the `kwargs` injection strategy, and takes a more explicit approach to determining which kwargs should be passed to `#initialize`'s super method for classes with an injector mixed in.

Before, we'd pass kwargs up to the super method according to these two rules:

1. Pass individual kwargs if they're named keyword arg on the super method
2. Pass _all_ kwargs if the super method contains a "keyrest" `**double_splat` arg

This cased problems with one particular use case: injecting extra dependencies into rom-rb `Changeset` objects, whose `#initialize` methods have a `#parameters` signature of `[[:rest, :args]]` (i.e. `*args`). rom-rb changesets extract keyword args via different means than named method parameters. The end result was that the `*args` parameter on the changeset initialize method meant that critical keywords args (like `__data__:`) were never being passed up to the changeset's own `#initialize` when we included an auto-injection mixin.

To handle this situation, I changed the handling of passing kwargs up to the super method to follow these rules:

1. If the keyword arg is not a named dependency (i.e. if its a keyword arg we're not explicitly dealing with for auto-injection), pass it up
2. If the super method has it as an explicitly named argument (even if it's also a dependency for auto-injection), pass it up too

Rule (1) above is what allows everything to get up to the rom-rb changeset `#initialize` as it should. Rule (2) is important for ensuring we still keep general compatibility with the broadest possible set of of keyword arg initializers.

This change passed all tests except for one. And the change I had to make there seemed eminently sensible to me (the previous behaviour didn't really make sense IMO).

What do you think, @solnic, @flash-gordon?

timriley commented:
(n.b. not sure why tests are failing on CI — I'll have to follow up on that failure)

timriley commented:
I’ll also add a test to exercise the example of a super initialize expecting *args and doing manual options extraction from there.

solnic commented:
LGTM but it would be good to add specs covering the case which we'll support

flash-gordon commented on spec/integration/args_injection_spec.rb:
```diff
@@ -9,7 +9,7 @@ module PassThroughInitializer
 
             def initialize(*)
-              super
+              super()
```
the original case was described in https://github.com/dry-rb/dry-auto_inject/pull/14#issuecomment-218562210 which was based on Hanami's code: https://github.com/hanami/controller/blob/858303b5cf2d158f0d34f37a58ad384cb4468e0f/lib/hanami/action/mime.rb#L172
It seems I can work around this using the args strategy

timriley replied:
@flash-gordon Ah, thanks for finding this. This is actually very interesting to revisit. That PR addressed the `args` strategy only, since at that point it was the default injection strategy. A couple of months later, with the 0.3.0 release (2016-06-02), the kwargs strategy became the default and the spec we committed in #14 continued to work purely because of internal implementation details of the method we generate in kwargs’ `#define_initialize_with_splat`. These details changed here in this PR, which is why this `args_injection_spec` broke.

I feel like we need to take a few steps here:

1. Restore the spec to what it was before this PR (using `super` instead of `super()`), but make it use the `args` strategy, as was the intention when the spec was written (`AutoInject = Dry::AutoInject({one: 1}).args`). I've tested this locally and it works, as you'd expect. This would be enough to get this current PR merged.
2. Add similar behaviour to the kwargs strategy, since it seems like it should be a helpful thing to incorporate in all our strategies, if possible.
3. Improve our descriptions of these specs so it's easier to understand their use cases simply by reading the spec files, now having to go back through PR history :)

What do you think? I'll commit the change for (1) right now, and if you're OK with this and all of my other changes today, we can get this PR merged.

flash-gordon commented on lib/dry/auto_inject/strategies/kwargs.rb:
```diff
@@ -44,24 +48,27 @@ def initialize(#{initialize_params})
           instance_mod.class_eval <<-RUBY, __FILE__, __LINE__ + 1
             def initialize(*args, **kwargs)
-              super(*args, #{super_kw_params})
+              dependency_names = [#{dependency_map.names.map { |name| ":#{name}" }.join(", ")}]
```
This now does much more on every `initialize`, could we instead assemble a static method body in a way it worked before?

timriley replied:
@flash-gordon Does https://github.com/dry-rb/dry-auto_inject/pull/41/commits/d042dd7707d38bfc4c1ae145880b75901a481de9 seem better to you?

timriley commented:
Thanks for the review @flash-gordon and @solnic. I’ll try make those performance improvements and add that other test case I mentioned before taking this further. Will ping you again then.

timriley commented on lib/dry/auto_inject/strategies/kwargs.rb:
```diff
@@ -14,7 +14,11 @@ def define_new
               }.merge(kwargs)
 
-              super(*args, **deps)
+              other_kwargs = kwargs.each_with_object({}) { |(key, val), hsh|
```
@flash-gordon much nicer!

timriley commented:
@solnic @flash-gordon This is all tidied up and ready for you if you want to take another look.

I've added a spec demonstrating support for the new behaviour we've added here. This spec also shows how I'd like to start organising our suite in the future (our single giant main spec file is getting out of hand).
